library/asn1write.c

```diff
@@ -101,10 +101,18 @@ int mbedtls_asn1_write_len( unsigned char **p, unsigned char *start, size_t len
 
 int mbedtls_asn1_write_tag( unsigned char **p, unsigned char *start, unsigned char tag )
 {
-    if( *p - start < 1 )
-        return( MBEDTLS_ERR_ASN1_BUF_TOO_SMALL );
+    return mbedtls_asn1_write_long_tag( p, start, tag );
+}
+
+int mbedtls_asn1_write_long_tag( unsigned char **p, unsigned char *start, int tag )
+{
+    while ( tag != 0 ) {
+        if( *p - start < 1 )
+            return( MBEDTLS_ERR_ASN1_BUF_TOO_SMALL );
 
-    *--(*p) = tag;
+        *--(*p) = tag & 0xFF;
+        tag >>= 8;
```

Comment:
Well, the original API doesn't check whether the tag that fits `unsigned char` is a valid tag that can be written on one byte. Following this scheme, I didn't add any validation or modification of the tag supplied. I agree that validating the tag is useful, but that would have nothing to do with what I'm adressing here and should be done in a seperate pull request.

Regarding your suggestion to use tags that are magically encoded with the correct flags is a bit more tricky:
1. Many specifications use the raw tags, which may lead to problems when bits get added magically
2. mbedtls doesn't provide any means to translate a raw tag into a human readable one. For example, in OpenSSL you could write a short raw tag `0x53` as `ASN1_APP_IMP(CVC_CHAT_SEQ, relative_authorization, ASN1_OCTET_STRING, 0x13)`. In mbedtls, however, there aren't any short cuts for the implicit application specific tag `0x13` so that the caller `mbedtls_asn1_write_tag` already has to translate this to the raw value `0x53`.

In summary, my implementation sticks to the current paradigms by choosing simplicity over too much magic.